### PR TITLE
Invoices

### PR DESCRIPTION
**Pull request by AlixH**
- The **download** of the invoice will be in the next PR

### CODE REVIEW

File: src/components/invoice/InvoiceComponent.tsx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  const [user, setUser] = useState<User>(null);
+
+  useEffect(() => {
+    setUp().catch((error) => {
+      console.error(I18n.t('invoices.invoicesUnexpectedError'), error);
+    });
+  });
+
+  async function setUp(): Promise<void> {
+    const centralServerProvider = await ProviderFactory.getProvider();
```
Comment (ClaudeROSSI) [inline src/components/invoice/InvoiceComponent.tsx:32]: I have a side-effect on my side. There is a kind of loop calling for ever the getUsers() method when displying the list of invoices.

While debugging, it seems the root cause is that useEffect() gets called again and again.
Can you reproduce it?

File: src/types/Server.tsx [hunk 2]
```diff
@@ -503,5 +504,8 @@ export enum ServerRoute {
   // BILLING URLs for Non-CRUD Operations on PAYMENT METHODS
   REST_BILLING_PAYMENT_METHOD_SETUP = 'users/:userID/payment-methods/setup',
   REST_BILLING_PAYMENT_METHOD_ATTACH = 'users/:userID/payment-methods/:paymentMethodID/attach',
-  REST_BILLING_PAYMENT_METHOD_DETACH = 'users/:userID/payment-methods/:paymentMethodID/detach'
+  REST_BILLING_PAYMENT_METHOD_DETACH = 'users/:userID/payment-methods/:paymentMethodID/detach',
+
+  // BILLING URLs for CRUD operations on INVOICES
+  REST_BILLING_INVOICES = 'invoices'
```
Comment (ClaudeROSSI) [inline src/types/Server.tsx:510]: Please also add the constants for downloading the invoice PDF

// BILLING URLs for CRUD operations on INVOICES
  REST_BILLING_INVOICES = 'invoices',
  REST_BILLING_INVOICE = 'invoices/:invoiceID',

  // BILLING URLs for Non-CRUD operations on INVOICES
  REST_BILLING_DOWNLOAD_INVOICE = 'invoices/:invoiceID/download',

File: src/types/Server.tsx [hunk 3]
```diff
@@ -398,6 +398,7 @@ export enum ServerAction {
   NOTIFICATIONS = 'Notifications',
 
   BILLING = 'Billing',
+  INVOICES = 'Invoices',
```
Comment (ClaudeROSSI) [inline src/types/Server.tsx:401]: ??? BILLING_INVOICES

**Review by ClaudeROSSI — CHANGES_REQUESTED** [review]
There is a kind of loop calling getUsers() forever.
Please check my comments

File: src/types/Server.tsx (hunk 2 above)
Comment (AlixH) [reply src/types/Server.tsx:510]: I will add it in the PR featuring the download

File: src/components/invoice/InvoiceComponent.tsx (hunk 1 above)
Comment (AlixH) [reply src/components/invoice/InvoiceComponent.tsx]: Yes, i have fixed it

File: src/types/Server.tsx (hunk 3 above)
Comment (AlixH) [reply src/types/Server.tsx]: This was unused anyway

File: src/components/invoice/InvoiceComponent.tsx [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+              </Text>
+            </View>
+          </View>
+          <View style={style.rightContainer}>
+            <View style={style.invoiceStatusContainer}>
+              <Chip statusStyle={[buildStatusStyle(invoice.status)]} text={buildStatus(invoice.status)} navigation={navigation} />
+            </View>
+            <View style={style.invoiceAmountContainer}>
+              <Text adjustsFontSizeToFit={true} numberOfLines={1} style={[style.text, style.invoiceAmount]}>
+                {I18nManager.formatCurrency(invoice.amount)}
```
Comment (ClaudeROSSI) [inline src/components/invoice/InvoiceComponent.tsx:110]: ACHTUNG! - Invoice amount are in cents! 

Please divide by 100 to get the correct price.
I think you could also take care here to consider the invoice currency while formatting.
Our web app is so far relying on: invoice.currency.toUpperCase()
Do you have access to it?

image

**Review by ClaudeROSSI — CHANGES_REQUESTED** [review]
amount are stored in cents in the DB!
Have a look to my comments.
Thanks

File: src/components/invoice/InvoiceComponent.tsx [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+import User from '../../types/User';
+import { DataResult } from '../../types/DataResult';
+import { Card, CardItem, Icon } from 'native-base';
+import Utils from '../../utils/Utils';
+
+export interface Props extends BaseProps {
+  invoice: BillingInvoice;
+}
+
+export default function InvoiceComponent(props: Props) {
```
Comment (LucasBrazi06) [inline src/components/invoice/InvoiceComponent.tsx:20]: I assume that you had no other choice to use the functional component

File: src/components/invoice/InvoiceComponent.tsx [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  const invoiceAmount = invoice.amount && invoice.amount / 100;
+
+  useEffect(() => {
+    async function setUp(): Promise<void> {
+      const centralServerProvider = await ProviderFactory.getProvider();
+      const result: DataResult<User> = await centralServerProvider.getUsers({ UserID: invoice.userID });
+      const fetchedUser = result?.result?.[0];
+      setUser(fetchedUser);
+    }
+    setUp().catch((error) => {
```
Comment (LucasBrazi06) [inline src/components/invoice/InvoiceComponent.tsx:33]: Use try...catch + async/await

File: src/components/invoice/InvoiceComponent.tsx [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+      console.error(I18n.t('invoices.invoicesUnexpectedError'), error);
+    });
+  }, [invoice.userID]);
+
+  function buildStatus(invoiceStatus: BillingInvoiceStatus): string {
```
Comment (LucasBrazi06) [inline src/components/invoice/InvoiceComponent.tsx:38]: Use arrow function. Same for all.

File: src/components/invoice/InvoiceComponent.tsx [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+              </Text>
+            </View>
+            <View style={style.invoiceDetailsContainer}>
+              <View style={style.transactionContainer}>
+                <Text numberOfLines={1} style={style.text}>
+                  {invoice.sessions?.length}
+                </Text>
+                {invoice.sessions && (
+                  <Text numberOfLines={1} style={style.text}>
+                    {' '}
```
Comment (LucasBrazi06) [inline src/components/invoice/InvoiceComponent.tsx:94]: Use a padding here, not a space.

File: src/theme/variables/commonColor.tsx [hunk 9]
```diff
@@ -109,6 +109,7 @@ export default {
   cardBorderColor: '#ccc',
   cardBorderRadius: 2,
   cardItemPadding: platform === PLATFORM.IOS ? 10 : 12,
+  cardShadowColor: '#000000',
```
Comment (LucasBrazi06) [inline src/theme/variables/commonColor.tsx:112]: Don't we have an existing shadow var?

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]

File: src/components/invoice/InvoiceComponent.tsx (hunk 4 above)
Comment (LucasBrazi06) [reply src/components/invoice/InvoiceComponent.tsx]: > ACHTUNG! - Invoice amount are in cents!

Discussed this with Claude and API should always returns Euros.

File: src/theme/variables/commonColor.tsx (hunk 9 above)
Comment (AlixH) [reply src/theme/variables/commonColor.tsx:112]: I did not find one

File: src/components/invoice/InvoiceComponent.tsx (hunk 5 above)
Comment (LucasBrazi06) [reply src/components/invoice/InvoiceComponent.tsx:20]: Go for the class then.
We'll do a full migration to the functional components at once when we will have no other choice.

Comment (AlixH) [reply src/components/invoice/InvoiceComponent.tsx:20]: ok

**Review by LucasBrazi06 — APPROVED** [review]

**Pull request merged**